### ISSUE TITLE
Segfault/stack smashing with tao

### ISSUE THREAD

**Issue by 2e0byo** [emacs-ng/emacs-ng#506]
**Describe the bug**
Emacs dies with either a segfault or a detected stack smash.  GDB isn't terribly reliable after stack overflow, but the problem seems to lie in display_string.  Sometimes we get a sigabrt from display_string trying to render off the lhs of the screen.

**To Reproduce**
```shell
./configure --with-winit=tao --with-wr-gl=gl --enable-rust-debug
```

Unfortunately the only way I am able to trigger this bug is by running spacemacs and pressing `SPC b b` (helm-switch-buffer) or `SPC SPC` (helm-meta-x).  I will bisect it but I think it's caused by the new font scaling code.  Since that code is likely to go away anyway this might just be a matter of waiting.  I'm not at all sure, however, how something is managing to affect emac's internal string handling.

**OS**
Arch/hyprland.

**Error callstack**
Ambiguous.  Frequently gdb just gives up with `() ??`.

**Comment by 2e0byo** [comment]
Hmm, I've just had helm fail with complaints about ^@.  Has one sneaked in somewehere?

**Comment by 2e0byo** [comment]
And now we have a sigabrt inside `wait_reading_process_output` explictly caused by `__stack_chk_fail`.  I think something is just stack smashing and the actual failure sites are random.

**Comment by declantsien** [comment]
2e0byo ***@***.***> writes:


> Ambiguous.  Frequently gdb just gives up with `() ??`. 

For the C backtrace, you need pass `CFLAGS='-O0 -g3'`. See `etc/DEBUG`
for more information.

**Comment by declantsien** [comment]
2e0byo ***@***.***> writes:

> And now we have a sigabrt inside `wait_reading_process_output` explictly caused by `__stack_chk_fail`.  I think something is just stack smashing and the actual failure sites are random.
>

Can you try with `--enable-winit-pselect`? I have this random before
using tokio async fd.

**Comment by 2e0byo** [comment]
yeah, that was next on the list.  But gdb can normally do a pretty good job of inferring the location even when not compiled with debug flags (at least in my limited experience).  When the stack-smashing detection kicks in it's fine, but when the stack overflows and causes some random segfault it's (currently) lost.

**Comment by 2e0byo** [comment]
> 2e0byo ***@***.***> writes:
> And now we have a sigabrt inside `wait_reading_process_output` explictly caused by `__stack_chk_fail`. I think something is just stack smashing and the actual failure sites are random.
> Can you try with `--enable-winit-pselect`? I have this random before using tokio async fd.

Definitely, that's a good shot.

I've got some more data on it: with tao the failure rate is 40% (out of 10 runs) and the first invocation of helm-M-x takes about a minute.  (After that everything in helm is snappy).  So if it's not the select loop it should just be a matter of profiling what on earth is taking so long.  (`helm-M-x` in a terminal takes a second or so.)  *Something* is pathological and occasionally killing the stack.

**Comment by 2e0byo** [comment]
Failure rates identical (40% and 50% out of 10 runs) between tao and winit.

Ofc winit passed the first 5 runs and tao failed the first 3 just to make me think it was significant :D

**Comment by 2e0byo** [comment]
This doesn't happen with `--enable-winit-pselect`.  I think at the very least we should add a warning to the docs?

Although it didn't happen before the recent merges, i.e. in my `dev` branch.

**Comment by 2e0byo** [comment]
Managed to profile a run with tokio:

![2023-03-03-13:00:14-screenshot](https://user-images.githubusercontent.com/22841038/222726462-410e6700-915e-47eb-ba09-695966144ed0.png)

Aside from why helm is even trying to load tramp just to get the *M-x* completions, I'm not much wiser---just blocking shouldn't kill the stack.

**Comment by 2e0byo** [comment]
> 2e0byo ***@***.***> writes:
> This doesn't happen with `--enable-winit-pselect`. I think at the very least we should add a warning to the docs?
> My bad. If your branch is based of my old tao branch, this pselect code is used by default. See the conversation from here[0]. [0] #479

Drat.  I saw that ofc at the time and did I quick check that everything behaved with tokio and then forgot about it.

This is sufficiently weird---I've still no real idea why it should happen; I thought the pselect/tokio loop was entirely inside rust-land? to be worth documenting I think.  Anyhow thank you very much!

**Comment by declantsien** [comment]
> I've still no real idea why it should happen; I thought the pselect/tokio loop was entirely inside rust-land?

We are using Rust ffi. Code is written in Rust doesn't help in this case. As for why it should happen, I have no clue either. I have no idea how the lisp interpreter works.

>to be worth documenting I think.

I added some documentation [here](https://github.com/emacs-ng/emacs-ng/blob/master/docs/webrender.md#random-crashes-with-winittao)
